src/core/layout/qgslayoutitemattributetable.cpp

```diff
@@ -693,6 +705,48 @@ QVariant QgsLayoutItemAttributeTable::replaceWrapChar( const QVariant &variant )
   return replaced;
 }
 
+#ifdef HAVE_SERVER_PYTHON_PLUGINS
+QgsLayoutTableColumns QgsLayoutItemAttributeTable::filteredColumns()
+{
+  QgsLayoutTableColumns allowedColumns { mColumns };
+
+  QgsVectorLayer *source { sourceLayer() };
+
+  if ( ! source )
+  {
+    return allowedColumns;
+  }
+
+  QSet<QString> allowedAttributes;
+  for ( const auto &c : qgis::as_const( allowedColumns ) )
+  {
+    if ( ! c.attribute().isEmpty() )
+      allowedAttributes.insert( c.attribute() );
```

Comment:
What about expression based columns?

Reply:
The intention was that they pass through untouched. I was convinced that the access control plugins did not handle expression fields but I'll double check it.

Reply:
isn't that a potential data leak though? If I had an expression which collated two string fields I'm not supposed to see, you wouldn't want that shown would you?

Reply:
I considered that, If I get this correctly, the `QgsLayoutTableColumn::mAttribute` is not properly named because it actually stores an expression in every case.

Passing the column header makes no sense in this context, so the best we can do here is to pass all `mAttribute`s to the plugin that will decide if the column is visible or not based on its logic.

An additional safety measure would probably be to filter out the forbidden attributes from the feature request directly (this is what Server WFS does).

Note that there are many other potential data leaks in the server (notably: label expression).

Reply:
I've added a test for exp column filters, I didn't need to change the code in any way and all columns `mAttribute`(s) are passed to the plugin including:
- expression columns
- "virtual" fields from the layer

The plugin can decide whether allow them or not so there is a way to hide them.

I think we are good to go with this implementation, I'm not sure if we need to add a lower level filtering at the QgsFeatureRequest level.

Reply:
@nyalldawson I've changed the logic to work on attributes names extracted from the expressions, this should be the right implementation.

I've also fixed atlas support.